File: src/whisper_typer/code_speaker/history.py
"""TTS history tracking and reporting for code_speaker."""

import json
import logging
from dataclasses import asdict, dataclass
from datetime import date, datetime
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)

HISTORY_DIR = Path.home() / ".code-speaker-history"


@dataclass
class TTSRecord:
    """Record of a single TTS event with per-stage latency."""
    timestamp: str
    event_type: str           # stop, permission, notification, manual
    input_text_chars: int     # Original text length
    summarized: bool          # Whether Ollama was used
    summary_text: str         # What was actually spoken
    ollama_latency_ms: int    # Summarization time (0 if not summarized)
    kokoro_latency_ms: int    # Audio generation time
    playback_duration_ms: int # Playback time
    total_latency_ms: int     # End-to-end TTS pipeline
    voice: str                # Kokoro voice used
    cancelled: bool = False
    claude_session_ms: Optional[int] = None  # Time Claude spent thinking
    reminder_count: int = 0   # Reminders fired before user responded


def save_tts_record(record: TTSRecord):
    """Save a TTS record to today's history file."""
    HISTORY_DIR.mkdir(parents=True, exist_ok=True)
    today = date.today().isoformat()
    filepath = HISTORY_DIR / f"{today}-tts.jsonl"

    with open(filepath, "a") as f:
        f.write(json.dumps(asdict(record)) + "\n")


def load_tts_records(target_date: str = "today") -> list[TTSRecord]:
    """Load TTS records for a given date."""
    if target_date == "today":
        target_date = date.today().isoformat()

    filepath = HISTORY_DIR / f"{target_date}-tts.jsonl"
    records = []

    if not filepath.exists():
        return records

    with open(filepath) as f:
        for line in f:
            try:
                data = json.loads(line)
                records.append(TTSRecord(**data))
            except (json.JSONDecodeError, TypeError):
                continue

    return records


def list_tts_dates() -> list[str]:
    """List dates with TTS history."""
    if not HISTORY_DIR.exists():
        return []

    dates = []
    for f in sorted(HISTORY_DIR.glob("*-tts.jsonl")):
        # Extract date from filename like "2026-02-12-tts.jsonl"
        name = f.stem.replace("-tts", "")
        dates.append(name)

    return dates


def generate_tts_report(target_date: str = "today") -> str:
    """Generate a TTS latency and usage report."""
    records = load_tts_records(target_date)

    if not records:
        return f"No TTS events recorded for {target_date}."

    if target_date == "today":
        target_date = date.today().isoformat()

    # Count by event type
    by_type = {}
    for r in records:
        by_type.setdefault(r.event_type, []).append(r)

    # Compute averages
    total = len(records)
    non_cancelled = [r for r in records if not r.cancelled]

    def avg(values):
        return sum(values) / len(values) if values else 0

    avg_ollama = avg([r.ollama_latency_ms for r in non_cancelled if r.summarized])
    avg_kokoro = avg([r.kokoro_latency_ms for r in non_cancelled])
    avg_playback = avg([r.playback_duration_ms for r in non_cancelled])
    avg_total = avg([r.total_latency_ms for r in non_cancelled])

    # Claude session duration (stop events only)
    claude_durations = [
        r.claude_session_ms for r in records
        if r.event_type == "stop" and r.claude_session_ms is not None
    ]
    avg_claude = avg(claude_durations)

    # Reminder stats
    reminder_counts = [r.reminder_count for r in records if r.reminder_count > 0]
    avg_reminders = avg(reminder_counts) if reminder_counts else 0

    # Build report
    lines = [
        f"=== Code Speaker TTS Report: {target_date} ===",
        "",
        f"Total events: {total}",
    ]

    for etype, recs in sorted(by_type.items()):
        lines.append(f"  {etype}: {len(recs)}")

    lines.extend([
        "",
        "--- Latency (non-cancelled events) ---",
        f"  Ollama summarize: {avg_ollama:.0f}ms avg"
        f" ({sum(1 for r in non_cancelled if r.summarized)} summarized)",
        f"  Kokoro generate:  {avg_kokoro:.0f}ms avg",
        f"  Playback:         {avg_playback:.0f}ms avg",
        f"  Total pipeline:   {avg_total:.0f}ms avg",
    ])

    if claude_durations:
        lines.extend([
            "",
            "--- Claude Session Duration ---",
            f"  Avg thinking time: {avg_claude/1000:.1f}s",
            f"  Events with timing: {len(claude_durations)}",
        ])

    if reminder_counts:
        lines.extend([
            "",
            "--- Reminders ---",
            f"  Events with reminders: {len(reminder_counts)}",
            f"  Avg reminders before response: {avg_reminders:.1f}",
            f"  Max reminders: {max(reminder_counts)}",
        ])

    # Cancelled events
    cancelled = [r for r in records if r.cancelled]
    if cancelled:
        lines.append(f"\nCancelled events: {len(cancelled)}")

    return "\n".join(lines)

File: whisper-typer-rs/src/code_speaker/history.rs
//! TTS event history and reporting.
//!
//! Stores TTS records in JSONL files at ~/.code-speaker-history/{date}-tts.jsonl.

use std::fs;
use std::io::Write;
use std::path::PathBuf;

use chrono::Local;
use serde::{Deserialize, Serialize};
use tracing::warn;

fn history_dir() -> PathBuf {
    dirs::home_dir()
        .expect("No home directory")
        .join(".code-speaker-history")
}

fn history_file(date: &str) -> PathBuf {
    history_dir().join(format!("{date}-tts.jsonl"))
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TTSRecord {
    pub timestamp: String,
    pub event_type: String,
    pub input_text_chars: usize,
    pub summarized: bool,
    pub summary_text: String,
    pub ollama_latency_ms: i64,
    pub kokoro_latency_ms: i64,
    pub playback_duration_ms: i64,
    pub total_latency_ms: i64,
    pub voice: String,
    pub cancelled: bool,
    pub reminder_count: u32,
}

pub fn save_tts_record(record: &TTSRecord) {
    let dir = history_dir();
    if let Err(e) = fs::create_dir_all(&dir) {
        warn!("Failed to create TTS history dir: {e}");
        return;
    }

    let date = Local::now().format("%Y-%m-%d").to_string();
    let path = history_file(&date);

    let mut file = match fs::OpenOptions::new().create(true).append(true).open(&path) {
        Ok(f) => f,
        Err(e) => {
            warn!("Failed to open TTS history file: {e}");
            return;
        }
    };

    match serde_json::to_string(record) {
        Ok(line) => {
            if let Err(e) = writeln!(file, "{line}") {
                warn!("Failed to write TTS history record: {e}");
            }
        }
        Err(e) => warn!("Failed to serialize TTS record: {e}"),
    }
}

pub fn load_tts_records(date: &str) -> Vec<TTSRecord> {
    let path = history_file(date);
    let contents = match fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    contents
        .lines()
        .filter_map(|line| serde_json::from_str(line).ok())
        .collect()
}

pub fn list_tts_dates() -> Vec<String> {
    let dir = history_dir();
    let entries = match fs::read_dir(&dir) {
        Ok(e) => e,
        Err(_) => return Vec::new(),
    };

    let mut dates: Vec<String> = entries
        .filter_map(|e| e.ok())
        .filter_map(|e| {
            let name = e.file_name().to_string_lossy().to_string();
            if name.ends_with("-tts.jsonl") {
                Some(name.trim_end_matches("-tts.jsonl").to_string())
            } else {
                None
            }
        })
        .collect();
    dates.sort();
    dates
}

pub fn generate_tts_report(date: &str) -> String {
    let records = load_tts_records(date);
    if records.is_empty() {
        return format!("No TTS records for {date}.");
    }

    let total = records.len();
    let cancelled = records.iter().filter(|r| r.cancelled).count();
    let summarized = records.iter().filter(|r| r.summarized).count();

    let avg_kokoro: f64 =
        records.iter().map(|r| r.kokoro_latency_ms as f64).sum::<f64>() / total as f64;
    let avg_total: f64 =
        records.iter().map(|r| r.total_latency_ms as f64).sum::<f64>() / total as f64;

    // Event type breakdown
    let mut event_counts = std::collections::HashMap::new();
    for r in &records {
        *event_counts.entry(r.event_type.as_str()).or_insert(0) += 1;
    }

    let mut report = format!(
        "# TTS Report for {date}\n\n\
        - Total events: {total}\n\
        - Cancelled: {cancelled}\n\
        - Summarized: {summarized}\n\
        - Avg Kokoro latency: {avg_kokoro:.0}ms\n\
        - Avg total latency: {avg_total:.0}ms\n\n\
        ## Event Types\n"
    );

    for (event, count) in &event_counts {
        report.push_str(&format!("- {event}: {count}\n"));
    }

    report
}
